server/src/main/java/io/spine/server/aggregate/Aggregate.java

```diff
@@ -392,19 +423,125 @@ protected String getMissingTxMessage() {
+     * 
+     * <p>The iterator is empty if there's no history for the aggregate.
+     *
+     * @param  criteria specifies the filters the returned data should pass and how the query 
+     *                  should be performed
+     * @return new iterator instance
+     */
+    public Iterator<AggregateEventRecord> historyBackward(AggregateRecordQueryCriteria criteria) {
+        checkNotNull(criteria);
+        final int batchSize = criteria.batchSize();
+        final List<AggregateEventRecord> items = queryBatch(batchSize);
+        return items.iterator();
+    }
+
+    /**
+     * Obtains a batch of aggregate records specified by the batch size.
+     * 
+     * @param batchSize the amount of events that must be retrieved
+     */
+    private List<AggregateEventRecord> queryBatch(int batchSize) {
+        checkPositive(batchSize);
+
+        final ImmutableList.Builder<AggregateEventRecord> builder = ImmutableList.builder();
+        int selectedCount = 0;
+
+        for (AggregateEventRecord record : historySinceLastSnapshot) {
+            builder.add(record);
+
+            selectedCount++;
+            if (selectedCount == batchSize) {
+                return builder.build();
+            }
+        }
+
         return builder.build();
     }
 
+    /**
+     * Creates a new list mapping every {@link Event} in the provided collection to a newly created 
+     * {@link AggregateEventRecord}.
+     */
+    private static List<AggregateEventRecord> toAggregateEventRecords(Collection<Event> events) {
+        final ImmutableList.Builder<AggregateEventRecord> items = ImmutableList.builder();
+        for (Event event : events) {
+            final AggregateEventRecord record = newAggregateEventRecord(event);
+            items.add(record);
+        }
+        return items.build();
+    }
+
+    /**
+     * Creates a new {@link AggregateEventRecord} holding the specified {@link Event}.
+     */
+    private static AggregateEventRecord newAggregateEventRecord(Event event) {
+        final EventContext context = event.getContext();
+        return AggregateEventRecord.newBuilder()
+                                   .setTimestamp(context.getTimestamp())
+                                   .setEvent(event)
+                                   .build();
+    }
+
+    /**
+     * Checks if the command was already handled by the aggregate since last snapshot.
+     *
+     * <p> The check is performed by searching for an event caused by this command that was
+     * committed since last snapshot.
+     *
+     * <p> This functionality support the ability to stop duplicate commands from being dispatched
+     * to the aggregate.
+     *
+     * @param command the command to check
+     * @return {@code true} if the command was handled since last snapshot, {@code false} otherwise
+     */
+    public boolean didHandleSinceLastSnapshot(CommandEnvelope command) {
```

Comment:
Let's have a separate (at least, inner) class called `IdempotencyGuard`. Its responsibility would be to provide idempotency for a certain type of input (now `Command`s, than later `Event`s and `Rejection`s to react on).